File: badapple.cpp
#include <opencv2/opencv.hpp>
#include <iostream>
#include <vector>
#include <chrono>
#include <thread>

void processFrame(const cv::Mat& frame, int blockSize, bool textMode, const cv::Mat& blackImg, const cv::Mat& whiteImg, cv::Mat& outputFrame) {
    int rows = frame.rows / blockSize;
    int cols = frame.cols / blockSize;

    cv::Mat resized;
    cv::resize(frame, resized, cv::Size(cols, rows));
    cv::cvtColor(resized, resized, cv::COLOR_BGR2GRAY);

    if (textMode) {
        std::cout << "\033[H"; // Move cursor to top-left
        for (int i = 0; i < rows; ++i) {
            for (int j = 0; j < cols; ++j) {
                uchar pixel = resized.at<uchar>(i, j);
                std::cout << (pixel < 128 ? '0' : '1') << (pixel < 128 ? '0' : '1');
            }
            std::cout << '\n';
        }
        std::cout.flush();
    } else {
        outputFrame = cv::Mat::zeros(frame.size(), frame.type());
        for (int i = 0; i < rows; ++i) {
            for (int j = 0; j < cols; ++j) {
                uchar pixel = resized.at<uchar>(i, j);
                cv::Mat roi = outputFrame(cv::Rect(j * blockSize, i * blockSize, blockSize, blockSize));
                if (pixel < 128) {
                    blackImg.copyTo(roi);
                } else {
                    whiteImg.copyTo(roi);
                }
            }
        }
    }
}

int main(int argc, char** argv) {
    bool textMode = false;
    std::vector<std::string> args;

    for (int i = 1; i < argc; ++i) {
        if (std::string(argv[i]) == "--text") {
            textMode = true;
        } else {
            args.push_back(argv[i]);
        }
    }

    if (args.size() < 2) {
        std::cout << "Usage:\n";
        std::cout << "  Text Mode (auto or forced):\n";
        std::cout << "    " << argv[0] << " <video_file> <block_size> [--text]\n";
        std::cout << "  Video Mode:\n";
        std::cout << "    " << argv[0] << " <video_file> <block_size> <black_img> <white_img> <output_video>\n";
        return -1;
    }

    if (!textMode && args.size() < 5) {
        textMode = true;
    }

    std::string videoPath = args[0];
    int blockSize = std::stoi(args[1]);

    std::string blackPixelImage, whitePixelImage, outputVideo;
    cv::Mat blackImg, whiteImg;

    if (!textMode) {
        blackPixelImage = args[2];
        whitePixelImage = args[3];
        outputVideo = args[4];
        blackImg = cv::imread(blackPixelImage);
        whiteImg = cv::imread(whitePixelImage);

        if (blackImg.empty() || whiteImg.empty()) {
            std::cerr << "Error loading replacement images." << std::endl;
            return -1;
        }
    }

    cv::VideoCapture cap(videoPath);
    if (!cap.isOpened()) {
        std::cerr << "Error opening video file." << std::endl;
        return -1;
    }

    cv::Mat frame, outputFrame;
    cv::VideoWriter videoWriter;

    double fps = cap.get(cv::CAP_PROP_FPS);
    int totalFrames = cap.get(cv::CAP_PROP_FRAME_COUNT);
    int currentFrame = 0;

    if (!textMode) {
        int fourcc = cv::VideoWriter::fourcc('a', 'v', 'c', '1');
        cv::Size frameSize(cap.get(cv::CAP_PROP_FRAME_WIDTH), cap.get(cv::CAP_PROP_FRAME_HEIGHT));
        videoWriter.open(outputVideo, fourcc, fps, frameSize);
        if (!videoWriter.isOpened()) {
            std::cerr << "Error: Could not open the output video file for writing." << std::endl;
            return -1;
        }
    }

    cv::setNumThreads(4);
    auto start_time = std::chrono::high_resolution_clock::now();

    if (textMode) std::cout << "\033[2J"; // Clear terminal once

    while (cap.read(frame)) {
        processFrame(frame, blockSize, textMode, blackImg, whiteImg, outputFrame);

        if (!textMode) {
            videoWriter.write(outputFrame);
        } else {
            std::this_thread::sleep_for(std::chrono::milliseconds(static_cast<int>(1000.0 / fps)));
        }

        currentFrame++;
        if (!textMode)
            std::cout << "\rProgress: Frame " << currentFrame << " / " << totalFrames << std::flush;
    }

    auto end_time = std::chrono::high_resolution_clock::now();
    std::cout << std::endl;

    std::chrono::duration<double> elapsed_time = end_time - start_time;
    std::cout << "Processing completed in " << elapsed_time.count() << " seconds." << std::endl;

    double avgfps = totalFrames / elapsed_time.count();
    double timeframe = elapsed_time.count() / totalFrames;
    std::cout << "Average Time per Frame: " << timeframe << " seconds. (" << avgfps << " fps)" << std::endl;

    return 0;
}

File: badapple.py
import cv2
import numpy as np
import sys
import time

import av

# pip install opencv-python
# pip install av

def process_frame(frame, block_size, black_img, white_img):
    rows, cols = frame.shape[0] // block_size, frame.shape[1] // block_size

    resized = cv2.resize(frame, (cols, rows))
    gray = cv2.cvtColor(resized, cv2.COLOR_BGR2GRAY)

    output_frame = np.zeros(frame.shape, dtype=frame.dtype)

    for i in range(rows):
        for j in range(cols):
            pixel = gray[i, j]
            roi = output_frame[i * block_size:(i + 1) * block_size, j * block_size:(j + 1) * block_size]
            if pixel < 128:
                roi[:] = black_img
            else:
                roi[:] = white_img
    return output_frame

def main():
    if len(sys.argv) < 6:
        print("Usage: python3 badapple.py <video_file> <block_size> <black_pixel_image> <white_pixel_image> <output_video>")
        sys.exit(-1)

    video_path = sys.argv[1]
    block_size = int(sys.argv[2])
    black_pixel_image = sys.argv[3]
    white_pixel_image = sys.argv[4]
    output_video = sys.argv[5]

    cap = cv2.VideoCapture(video_path)
    if not cap.isOpened():
        print("Error: Could not open video file.")
        sys.exit(-1)

    black_img = cv2.imread(black_pixel_image)
    white_img = cv2.imread(white_pixel_image)

    if black_img is None or white_img is None:
        print("Error: Could not load replacement images.")
        sys.exit(-1)

    frame_width = int(cap.get(cv2.CAP_PROP_FRAME_WIDTH))
    frame_height = int(cap.get(cv2.CAP_PROP_FRAME_HEIGHT))
    fps = cap.get(cv2.CAP_PROP_FPS)
    
    # Python version of openCV does not support native H.264 encoding
    # PyAV is better and will be used instead
    output_container = av.open(output_video, mode='w')
    stream = output_container.add_stream('libx264', rate=int(fps))
    stream.width = frame_width
    stream.height = frame_height
    stream.pix_fmt = 'yuv420p'  # Common pixel format for H.264 encoding
    
    total_frames = int(cap.get(cv2.CAP_PROP_FRAME_COUNT))
    current_frame = 0

    start_time = time.time()

    while True:
        ret, frame = cap.read()
        if not ret:
            break

        output_frame = process_frame(frame, block_size, black_img, white_img)
        #out.write(output_frame)
        
        # Convert frame to PyAV format
        frame_av = av.VideoFrame.from_ndarray(output_frame, format='bgr24')
        packet = stream.encode(frame_av)
        if packet:
            output_container.mux(packet)

        current_frame += 1
        print(f"\rProgress: Frame {current_frame} / {total_frames}", end="")
        
    # Flush the encoder
    packet = stream.encode(None)
    if packet:
        output_container.mux(packet)

    output_container.close()

    end_time = time.time()

    elapsed_time = end_time - start_time
    avg_fps = total_frames / elapsed_time
    time_per_frame = elapsed_time / total_frames
    print(f"\nProcessing completed in {elapsed_time:.5f} seconds.")
    print(f"Average Time per Frame: {time_per_frame:.9f} seconds. ({avg_fps:.2f} fps)")

    cap.release()
    #out.release()
    cv2.destroyAllWindows()

if __name__ == "__main__":
    main()